src/wasm-validator.h

```diff
@@ -46,13 +46,15 @@
 namespace wasm {
 
 // Print anything that can be streamed to an ostream
-template <typename T>
+  template <typename T,
+    typename std::enable_if<
+      !std::is_base_of<Expression, typename std::remove_pointer<T>::type>::value
```

Comment:
I think with something like that you'd have the same problem, i.e. you want it to be transparent to the caller (e.g.  a template like `fail`) whether you are dealing with an expr (in which case you want to wrap it with FullTypePrinter) or not (in which case you don't). So IIUC to do that you'd have to specialize/overload/templatize FullTypePrinter similarly.